docs/source/en/model_doc/clip.md

```diff
@@ -79,6 +79,123 @@ encode the text and prepare the images. The following example shows how to get t
+
+<Tip warning={true}>
+
+For small batch sizes, you might notice a slowdown in your model when using flash attention. Refer to the section [Expected speedups with Flash Attention and SDPA](#Expected-speedups-with-Flash-Attention-and-SDPA) below and select an appropriate attention implementation.
+
+</Tip>
+
+To load and run a model using Flash Attention 2, refer to the snippet below:
+
+```python
+>>> import torch
+>>> import requests
+>>> from PIL import Image
+
+>>> from transformers import CLIPProcessor, CLIPModel
+
+>>> device = "cuda"
+>>> torch_dtype = torch.float16
+
+>>> model = CLIPModel.from_pretrained(
+...     "openai/clip-vit-base-patch32",
+...     attn_implementation="flash_attention_2",
+...     device_map=device,
+...     torch_dtype=torch_dtype,
+... )
+>>> processor = CLIPProcessor.from_pretrained("openai/clip-vit-base-patch32")
+
+>>> url = "http://images.cocodataset.org/val2017/000000039769.jpg"
+>>> image = Image.open(requests.get(url, stream=True).raw)
+
+>>> inputs = processor(text=["a photo of a cat", "a photo of a dog"], images=image, return_tensors="pt", padding=True)
+>>> inputs.to(device)
+
+>>> with torch.no_grad():
+...     with torch.autocast(device):
+...         outputs = model(**inputs)
+
+>>> logits_per_image = outputs.logits_per_image  # this is the image-text similarity score
+>>> probs = logits_per_image.softmax(dim=1)  # we can take the softmax to get the label probabilities
+>>> print(probs)
+tensor([[0.9946, 0.0052]], device='cuda:0', dtype=torch.float16)
+```
+
+
+### Using Scaled Dot Product Attention (SDPA)
+
+PyTorch includes a native scaled dot-product attention (SDPA) operator as part of `torch.nn.functional`. This function 
+encompasses several implementations that can be applied depending on the inputs and the hardware in use. See the 
+[official documentation](https://pytorch.org/docs/stable/generated/torch.nn.functional.scaled_dot_product_attention.html) 
+or the [GPU Inference](https://huggingface.co/docs/transformers/main/en/perf_infer_gpu_one#pytorch-scaled-dot-product-attention)
+page for more information.
+
+SDPA is used by default for `torch>=2.1.1` when an implementation is available, but you may also set 
+`attn_implementation="sdpa"` in `from_pretrained()` to explicitly request SDPA to be used.
+
+```python
+from transformers import CLIPModel
+
+model = CLIPModel.from_pretrained("openai/clip-vit-base-patch32", torch_dtype=torch.float16, attn_implementation="sdpa")
+```
+
+For the best speedups, we recommend loading the model in half-precision (e.g. `torch.float16` or `torch.bfloat16`).
+
+### Expected speedups with Flash Attention and SDPA
+
+On a local benchmark (NVIDIA A10G, PyTorch 2.3.1+cu121) with `float16`, we saw the following speedups during inference for `"openai/clip-vit-large-patch14"` checkpoint ([code](https://gist.github.com/qubvel/ac691a54e54f9fae8144275f866a7ff8)):
```

Comment:
Just out of curiosity -- do these trends roughly hold for other checkpoints too? Anyway, you picked the most popular checkpoint I think since it's used in the diffusion community quite heavily.

Reply:
roughly hold, here is to "base" checkpoint

“openai/clip-vit-base-patch32”

CLIPTextModel
|   Num text labels |   Eager (s/iter) |   FA2 (s/iter) |   FA2 speedup |   SDPA (s/iter) |   SDPA speedup |
|------------------:|-----------------:|---------------:|--------------:|----------------:|---------------:|
|                 4 |            0.007 |          0.011 |         0.674 |           0.006 |          1.173 |
|                16 |            0.007 |          0.011 |         0.648 |           0.006 |          1.17  |
|                64 |            0.015 |          0.017 |         0.855 |           0.013 |          1.112 |
|               128 |            0.03  |          0.029 |         1.022 |           0.027 |          1.121 |
|               256 |            0.06  |          0.054 |         1.106 |           0.053 |          1.121 |

CLIPVisionModel
|   Image batch size |   Eager (s/iter) |   FA2 (s/iter) |   FA2 speedup |   SDPA (s/iter) |   SDPA speedup |
|-------------------:|-----------------:|---------------:|--------------:|----------------:|---------------:|
|                  1 |            0.007 |          0.005 |         1.224 |           0.005 |          1.307 |
|                  4 |            0.007 |          0.005 |         1.228 |           0.005 |          1.287 |
|                 16 |            0.007 |          0.006 |         1.23  |           0.005 |          1.316 |
|                 32 |            0.009 |          0.009 |         1.023 |           0.009 |          1.031 |
|                 64 |            0.018 |          0.017 |         1.076 |           0.017 |          1.082 |
|                128 |            0.035 |          0.032 |         1.09  |           0.032 |          1.095 |

CLIPModel
|   Image batch size |   Num text labels |   Eager (s/iter) |   FA2 (s/iter) |   FA2 speedup |   SDPA (s/iter) |   SDPA speedup |
|-------------------:|------------------:|-----------------:|---------------:|--------------:|----------------:|---------------:|
|                  1 |                 4 |            0.015 |          0.017 |         0.866 |           0.012 |          1.23  |
|                  1 |                16 |            0.015 |          0.017 |         0.86  |           0.012 |          1.221 |
|                  1 |                64 |            0.016 |          0.023 |         0.705 |           0.015 |          1.105 |
|                  4 |                 4 |            0.015 |          0.017 |         0.871 |           0.012 |          1.241 |
|                  4 |                16 |            0.015 |          0.017 |         0.866 |           0.012 |          1.231 |
|                  4 |                64 |            0.017 |          0.023 |         0.741 |           0.015 |          1.106 |
|                 16 |                 4 |            0.015 |          0.017 |         0.869 |           0.012 |          1.226 |
|                 16 |                16 |            0.015 |          0.017 |         0.878 |           0.012 |          1.234 |
|                 16 |                64 |            0.02  |          0.023 |         0.868 |           0.018 |          1.099 |
|                 32 |                 4 |            0.015 |          0.02  |         0.746 |           0.012 |          1.221 |
|                 32 |                16 |            0.015 |          0.02  |         0.756 |           0.013 |          1.215 |
|                 32 |                64 |            0.024 |          0.027 |         0.911 |           0.022 |          1.084 |

Reply:
FA2 results are super interesting!

Reply:
Okay, in general, SDPA seems to be still producing better speedups.